File: wordle.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

int find(char** word_list, char* word, int size_list){
    int found = 0;
    for (int i =0; i < size_list; i++){
        if(strcmp(word_list[i], word) == 0){
            found = 1;
            break;
        } 
    }

    return found;

}

int main(){
    char* list_of_words[] = {"about","above","abuse","actor","acute","admit","adopt","adult","after","again","agent","agree","ahead","alarm","album","alert","alike","alive","allow","alone","along","alter","among","anger","angle","angry","apart"};//,apple,apply,arena,argue,arise,array,aside,asset,audio,audit,avoid,award,aware,badly,baker,bases,basic,basis,beach,began,begin,begun,being,below,bench,billy,birth,black,blame,blind,block,blood,board,boost,booth,bound,brain,brand,bread,break,breed,brief,bring,broad,broke,brown,build,built,buyer,cable,calif,carry,catch,cause,chain,chair,chart,chase,cheap,check,chest,chief,child,china,chose,civil,claim,class,clean,clear,click,clock,close,coach,coast,could,count,court,cover,craft,crash,cream,crime,cross,crowd,crown,curve,cycle,daily,dance,dated,dealt,death,debut,delay,depth,doing,doubt,dozen,draft,drama,drawn,dream,dress,drill,drink,drive,drove,dying,eager,early,earth,eight,elite,empty,enemy,enjoy,enter,entry,equal,error,event,every,exact,exist,extra,faith,false,fault,fiber,field,fifth,fifty,fight,final,first,fixed,flash,fleet,floor,fluid,focus,force,forth,forty,forum,found,frame,frank,fraud,fresh,front,fruit,fully,funny,giant,given,glass,globe,going,grace,grade,grand,grant,grass,great,green,gross,group,grown,guard,guess,guest,guide,happy,harry,heart,heavy,hence,henry,horse,hotel,house,human,ideal,image,index,inner,input,issue,japan,jimmy,joint,jones,judge,known,label,large,laser,later,laugh,layer,learn,lease,least,leave,legal,level,lewis,light,limit,links,lives,local,logic,loose,lower,lucky,lunch,lying,magic,major,maker,march,maria,match,maybe,mayor,meant,media,metal,might,minor,minus,mixed,model,money,month,moral,motor,mount,mouse,mouth,movie,music,needs,never,newly,night,noise,north,noted,novel,nurse,occur,ocean,offer,often,order,other,ought,paint,panel,paper,party,peace,peter,phase,phone,photo,piece,pilot,pitch,place,plain,plane,plant,plate,point,pound,power,press,price,pride,prime,print,prior,prize,proof,proud,prove,queen,quick,quiet,quite,radio,raise,range,rapid,ratio,reach,ready,refer,right,rival,river,robin,roger,roman,rough,round,route,royal,rural,scale,scene,scope,score,sense,serve,seven,shall,shape,share,sharp,sheet,shelf,shell,shift,shirt,shock,shoot,short,shown,sight,since,sixth,sixty,sized,skill,sleep,slide,small,smart,smile,smith,smoke,solid,solve,sorry,sound,south,space,spare,speak,speed,spend,spent,split,spoke,sport,staff,stage,stake,stand,start,state,steam,steel,stick,still,stock,stone,stood,store,storm,story,strip,stuck,study,stuff,style,sugar,suite,super,sweet,table,taken,taste,taxes,teach,teeth,terry,texas,thank,theft,their,theme,there,these,thick,thing,think,third,those,three,threw,throw,tight,times,tired,title,today,topic,total,touch,tough,tower,track,trade,train,treat,trend,trial,tried,tries,truck,truly,trust,truth,twice,under,undue,union,unity,until,upper,upset,urban,usage,usual,valid,value,video,virus,visit,vital,voice,waste,watch,water,wheel,where,which,while,white,whole,whose,woman,women,world,worry,worse,worst,worth,would,wound,write,wrong"};
    int list_size = sizeof(list_of_words)/sizeof(list_of_words[0]);

    srand(time(NULL)); // Seed the random number generator to prevent it from always giving the same value
    int random = rand() % list_size; // Generate a random index within the array bounds

    char* word_chosen = list_of_words[random];
    printf("word: %s\n", word_chosen);
    char* word_entered = malloc(sizeof(char)*6);
    int tries = 0;
    char to_return[5] = "aaaaa";
    while ((word_chosen != word_entered) && tries < 5){
        printf("Enter your guess here: ");
        scanf("%5s", word_entered);  //only scan the first 5 characters
        if(strlen(word_entered) != 5 || find(list_of_words, word_entered, list_size) == 0){
            printf("Please ensure your guess is exactly 5 letters and a recognized word in the english dictionary\n");
        }
        else{
            if (word_entered[0] == word_chosen[0] && word_entered[1] == word_chosen[1] && word_entered[2] == word_chosen[2] && word_entered[3] == word_chosen[3] && word_entered[4] == word_chosen[4]){
                printf("Congratulations! You guessed %s correctly. It took you %d tries\n", word_chosen, tries+1);
                break;
    } 
            for (int i = 0; i < 5; i++){
                if(word_entered[i] == word_chosen[i]) {
                    char capitalized = word_chosen[i];
                    capitalized = capitalized -32;
                    to_return[i] = capitalized; // capitalized
                } else if (word_entered[i] == word_chosen[0] || word_entered[i] == word_chosen[1] || word_entered[i] == word_chosen[2] || word_entered[i] == word_chosen[3] || word_entered[i] == word_chosen[4] ){
                        to_return[i] = word_entered[i]; // lowercase
                        //add the case where the letter has already been marked as there and so shouldn't show up a second time
                }else { 
                        to_return[i] = '_';
                    }
                }
                printf("%s\n", to_return);
                tries++;
        }


    }

    if(word_chosen != word_entered && tries >= 5){
        printf("Sorry, you have no more tries left. The word was %s. Better luck next time!", word_chosen);
    }

    free(word_entered);

    return 0;
}

File: Wordle.py
'''
Wordle:
1. Computer selects a random 5 letter english word. Get the list for that online
2. The user guess the whole world. The user cannot guess a word bigger than or smaller than 5 letters, and cannot guess the same word again
3. if the word user guesses is not the same word that the computer generated, they lose 1 life per valid attempt. See point 2 for invalid attemtpts
4. Print the user input in a way where if a letter from ther input matches the word position for position, it will be capital, if it is there but at a different place, it will be lowercase,
and if it does not exist in the word, it will appear as a dash
5. If the words match or lives become 0, the game ends

attempt 1:

def display_word(secret_word, guessed_word):
    displayed_word = ""
    count = 0
    for subletters in guessed_word:
        for letters in secret_word:
            if subletters == letters and secret_word.find(subletters) == guessed_word.find(subletters):
                displayed_word += subletters.upper()
            elif subletters == letters and secret_word.find(subletters) != guessed_word.find(subletters):
                displayed_word += subletters.lower()
            else:
                count += 1
        if count == 5:
            displayed_word += "-"
        count = 0
    return displayed_word


attempt 2:

def display_word(secret_word, guessed_word):
    global displayed_word
    displayed_word = ""
    count = 0
    for subindexes in range(len(guessed_word)):  #issue   #2   #is
        for indexes in range(len(secret_word)):   #times  #4
            if secret_word[indexes] == guessed_word[subindexes] and indexes == subindexes:
                displayed_word += guessed_word[subindexes].upper()
                break
            elif secret_word[indexes] == guessed_word[subindexes] and indexes != subindexes:
                if guessed_word[subindexes] in secret_word and subindexes > indexes:
                    pass
                elif len([secret_word[indexes] for char in displayed_word]) == sum(char == secret_word[indexes] for char in secret_word):
                    pass
                else: 
                    displayed_word += guessed_word[subindexes].lower()
                    break
            else:
                count += 1
        if count == 5 or len(displayed_word) <= subindexes:
            displayed_word += "-"
        count = 0
    return displayed_word
'''

import random
from wordle_words import word_list
secret_word = random.choice(word_list).upper()


def wordle(secret_word):
    lives = 5
    tries = 1
    guessed_word = ""
    list_of_guesses = []
    print("Welcome to Wordle! \nThe goal of the game is to guess a five-letter english word. To start playing, enter a five letter word when prompted.\nRules: \n1. You have 5 tries\n2. A blank space means the letter is not in the word\n3. A lowercase letter indicates that the letter is in the word but not in the correct position]\n4. An uppercase letter means it is in the correct position\nEnjoy!")
    while lives > 0:
        guessed_word = input("Enter a 5 letter word: ").upper()
        if len(guessed_word) != 5 or guessed_word in list_of_guesses:
            print("Invalid input")
        else:
            list_of_guesses.append(guessed_word)
            if guessed_word == secret_word:
                print(f"Congratulations! {secret_word} is the right answer! It took you {tries} tries(y))")
                break
            else:
                lives -= 1
                tries += 1
                progress = display_word(secret_word, guessed_word)
                print(progress)
    if lives == 0:
        print(f"Youe died. The word was {secret_word}")
    

def display_word(secret_word, guessed_word):
    default_list = ["-" for i in range(5)]
    for subindex in range(len(guessed_word)):
        if guessed_word[subindex] in secret_word:       #ISSUE TIMES
            if guessed_word[subindex] == secret_word[subindex]:
                default_list[subindex] = guessed_word[subindex].upper()
            elif guessed_word[subindex] != secret_word[subindex]:
                default_list[subindex] = guessed_word[subindex].lower()
                if n_occurences(default_list, guessed_word[subindex].lower()) > n_occurences(secret_word,guessed_word[subindex]):
                    default_list[subindex] = "-"
    for char in secret_word:
        if n_occurences(default_list, char.lower()) + n_occurences(default_list, char.upper()) > n_occurences(secret_word, char):
            iterations_to_do = n_occurences(default_list, char.lower()) - (n_occurences(secret_word, char)-n_occurences(default_list, char.upper()))
            for i in range(iterations_to_do):
                default_list[default_list.index(char.lower())] = "-"
    stringed_version = "".join(str(e) for e in default_list)
    return stringed_version

def n_occurences(any_word, a_letter):
    counter = 0
    for letter in any_word:
        if letter == a_letter:
            counter += 1
    return counter


wordle(secret_word)